rust/hyperlane-base/src/contract_sync/cursor.rs

```diff
@@ -377,17 +382,26 @@ impl<T> ContractSyncCursor<T> for RateLimitedContractSyncCursor<T>
 where
     T: Send + Debug + 'static,
 {
-    async fn next_range(&mut self) -> ChainResult<(u32, u32, Duration)> {
-        self.rate_limit().await?;
+    async fn next_action(&mut self) -> ChainResult<(CursorAction, Duration)> {
         let to = u32::min(self.tip, self.from + self.chunk_size);
         let from = to.saturating_sub(self.chunk_size);
-        self.from = to + 1;
         let eta = if to < self.tip {
             self.eta_calculator.calculate(from, self.tip)
         } else {
             Duration::from_secs(0)
         };
-        Ok((from, to, eta))
+
+        let rate_limit = self.get_rate_limit().await?;
+        if let Some(rate_limit) = rate_limit {
+            return Ok((CursorAction::Sleep(rate_limit), eta));
+        } else {
+            self.from = to + 1;
```

Comment:
im a little confused on the relationship between `self.from` and the `from = to - chunk_size`
seems like we could maybe get away with having only `self.tip`